pricefeed/pricefeed.go

```diff
@@ -0,0 +1,143 @@
+
+var (
+	errUnsupportedNetwork = errors.New("unsupported network")
+	log                   = logging.Logger("pricefeed")
+)
+
+// PriceFeed contains the interesting data from a chainlink price feed query.
+type PriceFeed struct {
+	Description string // "COIN / USD"
+	Price       *apd.Decimal
+	UpdatedAt   time.Time
+}
+
+// GetETHUSDPrice returns the current ETH/USD price from the Chainlink oracle.
+// It errors if the chain ID is not the Ethereum mainnet.
+func GetETHUSDPrice(ctx context.Context, ec *ethclient.Client) (*PriceFeed, error) {
+	chainID, err := ec.ChainID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	switch chainID.Uint64() {
+	case common.MainnetChainID:
+		// No extra work to do
+	case common.GoerliChainID:
+		// Push stagenet/goerli users to a mainnet endpoint
+		ec, err = ethclient.Dial(mainnetEndpoint)
+		if err != nil {
+			return nil, err
+		}
+		defer ec.Close()
+	case common.GanacheChainID, common.HardhatChainID:
+		return &PriceFeed{
+			Description: "ETH / USD (fake)",
+			Price:       apd.New(123412345678, -8), // 1234.12345678
+			UpdatedAt:   time.Now(),
+		}, nil
+	default:
+		return nil, errUnsupportedNetwork
+	}
+
+	return getChainlinkPriceFeed(ctx, chainlinkETHToUSDProxy, ec)
+}
+
+// GetXMRUSDPrice returns the current XMR/USD price from the Chainlink oracle.
+// It errors if the chain ID is not the Ethereum mainnet.
+func GetXMRUSDPrice(ctx context.Context, ec *ethclient.Client) (*PriceFeed, error) {
+	chainID, err := ec.ChainID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	switch chainID.Uint64() {
+	case common.MainnetChainID:
+		// No extra work to do
+	case common.GoerliChainID:
+		// Push stagenet/goerli users to a mainnet endpoint
+		ec, err = ethclient.Dial(mainnetEndpoint)
+		if err != nil {
+			return nil, err
+		}
+		defer ec.Close()
+	case common.GanacheChainID, common.HardhatChainID:
+		return &PriceFeed{
+			Description: "XMR / USD (fake)",
+			Price:       apd.New(12312345678, -8), // 123.12345678
+			UpdatedAt:   time.Now(),
+		}, nil
+	default:
+		return nil, errUnsupportedNetwork
+	}
+
+	return getChainlinkPriceFeed(ctx, chainlinkXMRToUSDProxy, ec)
+}
+
+// getChainlinkPriceFeed retries the latest price feed data from the given contract address.
+func getChainlinkPriceFeed(ctx context.Context, feedAddress string, ec *ethclient.Client) (*PriceFeed, error) {
+
+	chainlinkPriceFeedProxy, err := contracts.NewAggregatorV3Interface(ethcommon.HexToAddress(feedAddress), ec)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := &bind.CallOpts{
+		Context: ctx,
+	}
+
+	roundData, err := chainlinkPriceFeedProxy.LatestRoundData(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	decimals, err := chainlinkPriceFeedProxy.Decimals(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	price := apd.NewWithBigInt(new(apd.BigInt).SetMathBigInt(roundData.Answer), -int32(decimals))
+	_, _ = price.Reduce(price) // push even multiples of 10 to the exponent
+	updatedAt := time.Unix(roundData.StartedAt.Int64(), 0)
```

Comment:
why `roundData.StartedAt` instead of `UpdatedAt`?

Reply:
They are identical, but you are right, that should have been UpdatedAt.  I'll fix it. Thanks.